tests/src/test/java/io/streamnative/pulsar/handlers/kop/SaslOauthKopHandlersTest.java

```diff
@@ -122,6 +134,33 @@ public void testSimpleProduceConsume() throws Exception {
         super.testSimpleProduceConsume();
     }
 
+    @Test(timeOut = 20000)
+    protected void testSimpleProduceConsumeWithTokenRefresh() throws Exception {
+        final String topic = "testSimpleProduceConsume";
+        final String message = "hello";
+
+        final Properties producerProps = newKafkaProducerProperties();
+        internalConfigureOauth2(producerProps, adminCredentialPath, CustomOauthLoginCallbackHandler.class);
+        @Cleanup final KafkaProducer<String, String> producer = new KafkaProducer<>(producerProps);
+
+        // Check that a token has been generated
+        assertEquals(1, CustomOauthLoginCallbackHandler.tokens.size());
+        OAuthBearerToken originalToken = CustomOauthLoginCallbackHandler.tokens.get(0);
+
+        // Sleep until the original OAuth token expires
+        Thread.sleep(originalToken.lifetimeMs() - System.currentTimeMillis());
+
+        // Check that new tokens have been generated
+        assertTrue(CustomOauthLoginCallbackHandler.tokens.size() > 1);
+
+        producer.send(new ProducerRecord<>(topic, message)).get();
+        RecordMetadata metadata = producer.send(new ProducerRecord<>(topic, message)).get();
+
+        log.info("Send to {}-partition-{}@{}", metadata.topic(), metadata.partition(), metadata.offset());
+        log.error("tokens: " + CustomOauthLoginCallbackHandler.tokens.stream().map(Object::toString)
+                .collect(Collectors.joining()));
```

Comment:
Why do you print error logs here? I think we should verify the size of `tokens` is 2 and `tokens[1]` is different from `tokens[0]`.

Reply:
Oops, forgot to remove debug logs - I didn't verify it's 2 so that if the test environment somehow gets very slow and the token refreshed more than once there won't be a false positive (checking there's only one token in the beginning seems less risky), but I guess we can start with 2 and see how it goes.